File: server/aitutor_server/ocr/lang_detect.py
"""Language detection over OCR'd text. Trivial Unicode-range heuristic — works because
PSLE essays are exclusively English or Simplified Chinese (no other scripts in scope)."""
from __future__ import annotations

from aitutor_server.api.schemas import Language

_CJK_RANGES = (
    (0x4E00, 0x9FFF),    # CJK Unified Ideographs
    (0x3400, 0x4DBF),    # CJK Unified Ideographs Extension A
    (0xF900, 0xFAFF),    # CJK Compatibility Ideographs
    (0xFF00, 0xFFEF),    # Halfwidth/fullwidth forms (commonly mixed in)
)


def _is_cjk(cp: int) -> bool:
    return any(lo <= cp <= hi for lo, hi in _CJK_RANGES)


def detect_language(text: str) -> Language:
    """Return ``"zh-Hans"`` if ≥10% of letter-ish characters are CJK, else ``"en"``."""
    if not text:
        return "en"
    cjk = sum(1 for ch in text if _is_cjk(ord(ch)))
    letters = sum(1 for ch in text if ch.isalpha() or _is_cjk(ord(ch)))
    if letters == 0:
        return "en"
    return "zh-Hans" if (cjk / letters) >= 0.10 else "en"

File: android/core/src/main/kotlin/com/aitutor/core/LangDetect.kt
package com.aitutor.core

// Port of server/aitutor_server/lang_detect.py. Trivial Unicode-range heuristic
// -- works because PSLE essays are exclusively English or Simplified Chinese.

object LangDetect {
    private val CJK_RANGES = listOf(
        0x4E00..0x9FFF, // CJK Unified Ideographs
        0x3400..0x4DBF, // Extension A
        0xF900..0xFAFF, // Compatibility Ideographs
        0xFF00..0xFFEF, // Half/fullwidth forms
    )

    private fun isCjk(cp: Int): Boolean = CJK_RANGES.any { cp in it }

    /** "zh-Hans" if >= 10% of letter-ish characters are CJK, else "en". */
    fun detect(text: String): String {
        if (text.isEmpty()) return "en"
        var cjk = 0
        var letters = 0
        for (ch in text) {
            val c = isCjk(ch.code)
            if (c) cjk++
            if (ch.isLetter() || c) letters++
        }
        if (letters == 0) return "en"
        return if (cjk.toDouble() / letters >= 0.10) "zh-Hans" else "en"
    }
}
